SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
++    {
++        // print to file
++        fprintf(pTracking->pbOutfile, "%s", pszLintedCharData);
++    }
++    else if (pTracking->nInPackage && !pTracking->nTimeFound)
++    {
++        // add to buffer
++        checkAndResizeBuffer(&(pTracking->pszElementBuffer), &(pTracking->nBufferMaxLen), length);
++        strcat(pTracking->pszElementBuffer, pszLintedCharData);
++    } // else do nothing (skipped package)
++
++cleanup:
++    if (pszLintedCharData)
++    {
++        free(pszLintedCharData);
++    }
++    if (pszCharData)
++    {
++        free(pszCharData);
++    }
++    return;
++error:
++    goto cleanup;
++}
++
++char * 
++TDNFFilterFile(const char * pszInFilePath, char * pszEpochTime) {
++    // vars
++    uint32_t dwError = 0;
++    TDNFFilterData * pData = NULL;
++    FILE * pbInFile = NULL;
++    FILE * pbOutFile = NULL;
++    XML_Parser bParser;
++    bzero(&bParser, sizeof(XML_Parser));
++    char pszTimeExtension[100];
++    char * pszOutFilePath = NULL;
++    time_t nEpochTime;
++    bzero(&nEpochTime, sizeof(time_t));
++    char * pszEpochEnd = NULL;
++    nEpochTime = strtoll(pszEpochTime, &pszEpochEnd, 10);
++    
++    if (nEpochTime == 0 && strcmp(pszEpochTime, "0") != 0) {
++        printf("Error converting --epochtimelimit: %s\n", pszEpochTime);
++        dwError = ERROR_TDNF_TIME_FILTER_PARSE;
++        BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++    }
++    if (pszEpochEnd != NULL && strcmp(pszEpochEnd, "") != 0) {
++        printf("The time provided to --epochtimelimit is misformatted: \'%s\' should be an epoch time\n", pszEpochTime);
++        dwError = ERROR_TDNF_TIME_FILTER_PARSE;
++        BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++    }
++
++    sprintf(pszTimeExtension, "-%ld.xml", nEpochTime);
++    int nOutFileLen = strlen(pszInFilePath) + strlen(pszTimeExtension);
++
++    pszOutFilePath = malloc(nOutFileLen * sizeof(char));
++    if (!pszOutFilePath)
++    {
++        dwError = ERROR_TDNF_TIME_FILTER_MEMORY;
++        BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++    }
++    bzero(pszOutFilePath, nOutFileLen * sizeof(char));
++
++    // find total extension length
++    int nFileExtLen = 4; // len of ".xml"
++    char * pszFileExt = strrchr(pszInFilePath, '.');
++    if (strcmp(pszFileExt, ".xml") != 0)
++    {
++        nFileExtLen += strlen(pszFileExt);
++    }
++
++    strncpy(pszOutFilePath, pszInFilePath, strlen(pszInFilePath) - nFileExtLen); // remove extension to be added with the name
++    strcat(pszOutFilePath, pszTimeExtension);
++
++
++    // init vars, load files
++    pbInFile = solv_xfopen(pszInFilePath, "r");
++    if (!pbInFile)
++    {
++        dwError = ERROR_TDNF_TIME_FILTER_IO;
++        BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++    }
++    pbOutFile = fopen(pszOutFilePath, "w");
++    if (!pbOutFile)
++    {
++        dwError = ERROR_TDNF_TIME_FILTER_IO;
++        BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError);
++    }
++
++    pData = (TDNFFilterData *) malloc(sizeof(TDNFFilterData));
```

Comment:
Does this need to be allocated on the Heap?

Reply:
Not inherently required. This was an artifact of the prototype design prior to using heap space for the element buffer. Will change to use a stack var.